proposals/0000-ghc2023.rst

```diff
@@ -0,0 +1,116 @@
+
+The first edition, ``GHC2021``, is now two years old, and I propose we define
+``GHC2023`` now, even if turns out to be a modest change, to keep the process alive.
+
+Proposed Change Specification
+-----------------------------
+
+The ``GHC2023`` language extension set comprises of all language extensions
+that ``GHC2021`` comprises, plus the following
+
+* `ExplicitNamespaces`_
+* `LambdaCase`_
+
+
+(*The committe will vote on each extension separately. Until then, this is the
+list of nominated extensions, and will be updated after a vote.*)
+
+Why GHC2023 in the first place
+------------------------------
+There were concerns that defining ``GHC2023`` is too soon, given that
+``GHC2021`` has hardly reached the target audience, and that this introduces
+instability.
+
+I believe that neither are good reasons to not define ``GHC2023``:
+
+* There is an inherent latency between defining language editions and them
+reaching the users. It reaches users who do not have to worry about supporting
+older GHC first, and thus library authors later. This is somewhat unavoidable,
+but not per se a reason to reduce the frequency.
+
+* Relatively frequent language editions do not introduce undue instability:
+Users who pin the language edition in their `.cabal` file or so are not
+affected by the existence of a new one. Neither are users who pin their GHC
+version. Only those users who upgrade their version of GHC _and_ ask for the
+latest edition may now have access to new features.
+.
+As `Arnaud explained on the mailing list <https://mail.haskell.org/pipermail/ghc-steering-committee/2022-November/002949.html>`_, Rust (generally not perceived as a language with a perception of low stability) has a very similar model with their “language editions”.
+
+Furthermore, a regular, not too slow cadence makes the whole process more
+predictable, which I expect *improves* the perception of stability, in the
+sense of delivering a stable flow of changes.
+
+
+Why add ExplicitNamespaces?
+---------------------------
+
+As discussed in `issue #551
+<https://github.com/ghc-proposals/ghc-proposals/issues/551>`_, ``GHC2021``
+includes `TypeOperators`_, but does not include `ExplicitNamespaces`_. This was
+a very strange (and probably inadvertent) decision, given that the flag
+``-XTypeOperators`` enables both the `TypeOperators`_ and the
+`ExplicitNamespaces`_ language extension.
+
+It seems to be a bad idea to retroactive *change* ``GHC2021`` to fix this (we
+do care about stability after all), but we should certainly fix this in the
+upcoming edition.
+
+And – at least if one accepts that regular releases of ``GHC20xx`` are a Good
+Thing™, adding `ExplicitNamespaces`_ alone should be sufficient to cut a new
+release.
+
+A counter-argument to `ExplicitNamespaces`_ is that it has seen changes
+recently, e.g. in `#281 <https://github.com/ghc-proposals/ghc-proposals/pull/302>_` and
```

Comment:
```suggestion
A counter-argument to `ExplicitNamespaces`_ is that it has seen changes
recently, e.g. in `#281 <https://github.com/ghc-proposals/ghc-proposals/pull/281>`_ and
```

Incidentally, I find it very surprising that #281 modified `ExplicitNamespaces`. We should add new features to new extensions rather than old ones!

Reply:
I am not actually sure. I think the main reason we have extensions in the first place is that we had thing thing called the report that was relatively sacrosanct. If we did not have that, would we even have started this whole business?

Also, I think of language extension flags a bit like library imports: it's a bit of red tape you put on top of the file to unlock certain features. Not so different from `import`! And, like there, some things are imported by default, somethings not…

Would you say that we should stop extending or changing libraries, and create new ones whenever we evolve a feature? Probably not.

Reply:
Language extensions are more like importing a name than importing a module and yes, to export more features one should export more names, not change the behaviour of existing exported names, typically.

Reply:
Tom, I'm not sure why you feel this way. I would certainly not be able to make such a confident statement (though I would rather tend to prefer Joachim's metaphor).

At any rate, it's generally considered perfectly fine to generalise the type of a function in subsequent versions of a library as it's backward compatible (though it's not forward compatible). So even in your metaphor, it would seem justified to change the meaning of an extension in a backward compatible way.

Rust editions, which are featured in the motivation section of the proposal do, as far as I know, incorporate backward compatible changes too with new compiler versions. In this mindset, it would be ok to have an extension change in a backward compatible way even if it is part of a `GHC20XX` extension set.

Reply:
Naturally it's hard to convey the fully subtlety of a breakage policy in a single sentence. I've written [a much more detailed exploration elsewhere](http://h2.jaguarpaw.co.uk/posts/opaleyes-api-breakage-policy/). I was just surprised by the analogy that Joachim drew. I think libraries should typically evolve  by adding *new* names, not changing the functionality of old ones, and I believe the same for language extensions.

Reply:
I see. Thanks for the extra explanation: it's helpful. I don't think I fully agree, but I see where you're coming from now.